src/ucs/sys/topo.c

```diff
@@ -19,9 +42,85 @@ ucs_sys_bus_id_t ucs_sys_bus_id_unknown = { .domain   = 0xffff,
                                             .function = 0xff
                                           };
 
+static ucs_bus_id_bit_rep_t ucs_topo_get_bus_id_bit_repr(const ucs_sys_bus_id_t *bus_id)
+{
+    return (((uint64_t)bus_id->domain << 24) |
+            ((uint32_t)bus_id->bus << 16)    |
+            ((uint16_t)bus_id->slot << 8)    |
+            (bus_id->function));
+}
+
+void ucs_topo_init()
+{
+    ucs_spinlock_init(&ucs_topo_ctx.common_lock, 0);
+    kh_init_inplace(bus_to_sys_dev, &ucs_topo_ctx.bus_to_sys_dev_hash);
+    ucs_topo_ctx.sys_dev_to_bus_lookup.count = 0;
+}
+
+void ucs_topo_cleanup()
+{
+    ucs_status_t status;
+
+    kh_destroy_inplace(bus_to_sys_dev, &ucs_topo_ctx.bus_to_sys_dev_hash);
+
+    status = ucs_spinlock_destroy(&ucs_topo_ctx.common_lock);
+    if (status != UCS_OK) {
+        ucs_warn("ucs_recursive_spinlock_destroy() failed: %s",
+                 ucs_status_string(status));
+    }
+}
+
+static int ucs_topo_compare_bus_id(const ucs_sys_bus_id_t *bus_id1, 
+                                   const ucs_sys_bus_id_t *bus_id2)
+{
+    if ((bus_id1->domain == bus_id2->domain) &&
+        (bus_id1->bus == bus_id2->bus) &&
+        (bus_id1->slot == bus_id2->slot) &&
+        (bus_id1->function == bus_id2->function)) {
+        return 1;
+    }
+
+    return 0;
+}
+
 ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
                                             ucs_sys_device_t *sys_dev)
 {
+    khiter_t hash_it;
+    ucs_kh_put_t kh_put_status;
+
+    *sys_dev = UCS_TOPO_DEV_ID_UNKNOWN;
+
+    if (ucs_topo_compare_bus_id(bus_id, &ucs_sys_bus_id_unknown)) {
+        ucs_debug("found unknown device index %u for bus id %ld",
+                  *sys_dev, ucs_topo_get_bus_id_bit_repr(bus_id));
```

Comment:
calculate ucs_topo_get_bus_id_bit_repr(bus_id) once, in the beginning of the function